cmd/abapAddonAssemblyKitCheckCVs.go

```diff
@@ -0,0 +1,103 @@
+	// reroute command output to logging framework
+	c.Stdout(log.Writer())
+	c.Stderr(log.Writer())
+
+	client := piperhttp.Client{}
+
+	// error situations should stop execution through log.Entry().Fatal() call which leads to an os.Exit(1) in the end
+	err := runAbapAddonAssemblyKitCheckCVs(&config, telemetryData, &client, cpe, abaputils.ReadAddonDescriptor)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+
+func runAbapAddonAssemblyKitCheckCVs(config *abapAddonAssemblyKitCheckCVsOptions, telemetryData *telemetry.CustomData, client piperhttp.Sender,
+	cpe *abapAddonAssemblyKitCheckCVsCommonPipelineEnvironment, readAdoDescriptor abaputils.ReadAddonDescriptorType) error {
+	var addonDescriptorFromCPE abaputils.AddonDescriptor
+	json.Unmarshal([]byte(config.AddonDescriptor), &addonDescriptorFromCPE)
+	addonDescriptor, err := readAdoDescriptor(config.AddonDescriptorFileName)
+	if err != nil {
+		return err
+	}
+	addonDescriptor = combineYAMLRepositoriesWithCPEProduct(addonDescriptor, addonDescriptorFromCPE)
+	conn := new(abapbuild.Connector)
+	conn.InitAAKaaS(config.AbapAddonAssemblyKitEndpoint, config.Username, config.Password, client)
+
+	for i := range addonDescriptor.Repositories {
+		var c cv
+		c.initCV(addonDescriptor.Repositories[i], *conn)
+		err := c.validate()
+		if err != nil {
+			return err
+		}
+		c.copyFieldsToRepo(&addonDescriptor.Repositories[i])
+	}
+	log.Entry().Info("Write the resolved versions to the CommonPipelineEnvironment")
+	toCPE, _ := json.Marshal(addonDescriptor)
+	cpe.abap.addonDescriptor = string(toCPE)
+	return nil
+}
+
+//take the product part from CPE and the repositories part from the YAML file
+func combineYAMLRepositoriesWithCPEProduct(addonDescriptor abaputils.AddonDescriptor, addonDescriptorFromCPE abaputils.AddonDescriptor) abaputils.AddonDescriptor {
+	addonDescriptorFromCPE.Repositories = addonDescriptor.Repositories
+	return addonDescriptorFromCPE
+}
+
+func (c *cv) initCV(repo abaputils.Repository, conn abapbuild.Connector) {
+	c.Connector = conn
+	c.Name = repo.Name
+	c.VersionYAML = repo.VersionYAML
+}
+
+func (c *cv) copyFieldsToRepo(initialRepo *abaputils.Repository) {
+	initialRepo.Version = c.Version
+	initialRepo.SpLevel = c.SpLevel
+	initialRepo.PatchLevel = c.PatchLevel
+}
+
+func (c *cv) validate() error {
+	log.Entry().Infof("Validate component %s version %s and resolve version", c.Name, c.VersionYAML)
+	appendum := "/odata/aas_ocs_package/ValidateComponentVersion?Name='" + c.Name + "'&Version='" + c.VersionYAML + "'"
+	body, err := c.Connector.Get(appendum)
+	if err != nil {
+		return err
+	}
+	var jCV jsonCV
+	json.Unmarshal(body, &jCV)
+	c.Name = jCV.CV.Name
+	c.Version = jCV.CV.Version
+	c.SpLevel = jCV.CV.SpLevel
+	c.PatchLevel = jCV.CV.PatchLevel
+	log.Entry().Infof("Resolved version %s, splevel %s, patchlevel %s", c.Version, c.SpLevel, c.PatchLevel)
+	return nil
+}
+
+type jsonCV struct {
+	CV *cv `json:"d"`
+}
+
+type cv struct {
```

Comment:
Is it worth extracting `cv` and it's functions also to the `abap` package?

Reply:
BTW: what does `cv` stand for?

Reply:
cv is the acronym for "(software)component version" which are the (more or less) independent deployable units of an ABAP system. 
Honestly I'm new to go and do not know where to place the coding best. As far as I got the cv struct is only used in this step without reuse in any of the others - that's why I have not extracted it. Anyway what I read so far about recommended directory structure I was a bit surprised there is no "internal" package. So what's the suggestion in this case? Should it be extracted even if there is only on step using it?